lib/tests/test_revset.rs

```diff
@@ -1556,6 +1556,147 @@ fn test_evaluate_expression_connected() {
+        .write()
+        .unwrap();
+    let graph3commit6 = create_random_commit(mut_repo, &settings)
+        .set_parents(vec![graph3commit3.id().clone()])
+        .write()
+        .unwrap();
+    let graph3commit7 = create_random_commit(mut_repo, &settings)
+        .set_parents(vec![graph3commit4.id().clone(), graph3commit5.id().clone()])
+        .write()
+        .unwrap();
+
+    // Boundary heads are respected.
+    assert_eq!(
+        resolve_commit_ids(
+            mut_repo,
+            &format!(
+                "reachable({}, {})",
+                graph1commit2.id().hex(),
+                graph1commit1.id().hex()
+            )
+        ),
+        vec![graph1commit3.id().clone(), graph1commit2.id().clone(),]
+    );
+    assert_eq!(
+        resolve_commit_ids(
+            mut_repo,
+            &format!(
+                "reachable({}, {})",
+                graph1commit2.id().hex(),
+                graph1commit3.id().hex()
+            )
+        ),
+        vec![]
+    );
+
+    // Each graph is identifiable from any node in it.
+    for (i, commit) in [&graph1commit1, &graph1commit2, &graph1commit3]
+        .iter()
+        .enumerate()
+    {
+        assert_eq!(
+            resolve_commit_ids(
+                mut_repo,
+                &format!("reachable({}, root())", commit.id().hex())
+            ),
+            vec![
+                graph1commit3.id().clone(),
+                graph1commit2.id().clone(),
+                graph1commit1.id().clone(),
+            ],
+            "commit {}",
+            i + 1
+        );
+    }
+
+    for (i, commit) in [&graph2commit1, &graph2commit2, &graph2commit3]
+        .iter()
+        .enumerate()
+    {
+        assert_eq!(
+            resolve_commit_ids(
+                mut_repo,
+                &format!("reachable({}, root())", commit.id().hex())
+            ),
+            vec![
+                graph2commit3.id().clone(),
+                graph2commit2.id().clone(),
+                graph2commit1.id().clone(),
+            ],
+            "commit {}",
+            i + 1
+        );
+    }
+
+    for (i, commit) in [
+        &graph3commit1,
+        &graph3commit2,
+        &graph3commit3,
+        &graph3commit4,
+        &graph3commit5,
+        &graph3commit6,
+        &graph3commit7,
+    ]
+    .iter()
+    .enumerate()
+    {
+        assert_eq!(
+            resolve_commit_ids(
+                mut_repo,
+                &format!("reachable({}, root())", commit.id().hex())
+            ),
+            vec![
+                graph3commit7.id().clone(),
+                graph3commit6.id().clone(),
+                graph3commit5.id().clone(),
+                graph3commit4.id().clone(),
+                graph3commit3.id().clone(),
+                graph3commit2.id().clone(),
+                graph3commit1.id().clone(),
+            ],
+            "commit {}",
+            i + 1
+        );
+    }
+}
```

Comment:
Some of these test cases might be interesting:
* start/exclude from middle: `reachable(g3c4, g3c5)`
* start not in domain

Reply:
Second one is already covered, added a test for the first one.